File: tetris-app/backend/controller/game_controller.py
import random
from model.pieces import PIECES, MINOS, SRS_JLSTZ, SRS_I, GRID_WIDTH, GRID_HEIGHT, NUM_NEXT_PIECES
from model.game_state import GameState


class GameController:
    def __init__(self):
        self.state = GameState()
        self._refill_next()
        self._pop_next_piece()

    # --- internal helpers ---

    def _refill_next(self):
        if len(self.state.next_pieces) <= NUM_NEXT_PIECES:
            bag = list(PIECES)
            random.shuffle(bag)
            self.state.next_pieces += bag

    def _pop_next_piece(self):
        self._refill_next()
        self.state.rot = 0
        self.state.current_piece_shape = self.state.next_pieces.pop(0)
        self.state.current_piece = MINOS[self.state.current_piece_shape][0]
        self.state.current_piece_position = [0, GRID_WIDTH // 2 - 2]
        self._update_ghost()
        if not self._is_valid_position():
            self.state.game_over = True

    def _is_valid_position(self, piece=None, pos=None):
        if piece is None:
            piece = self.state.current_piece
        if pos is None:
            pos = self.state.current_piece_position
        for y, row in enumerate(piece):
            for x, cell in enumerate(row):
                if cell:
                    nx = pos[1] + x
                    ny = pos[0] + y
                    if nx < 0 or nx >= GRID_WIDTH or ny >= GRID_HEIGHT:
                        return False
                    if ny >= 0 and self.state.board[ny][nx]:
                        return False
        return True

    def _update_ghost(self):
        pos = self.state.current_piece_position[:]
        while self._is_valid_position(self.state.current_piece, pos):
            pos[0] += 1
        pos[0] -= 1
        self.state.ghost_position = pos

    def _lock_piece(self):
        shape = self.state.current_piece_shape
        pos = self.state.current_piece_position
        for y, row in enumerate(self.state.current_piece):
            for x, cell in enumerate(row):
                if cell:
                    self.state.board[pos[0] + y][pos[1] + x] = shape

    def _clear_lines(self):
        new_board = [row for row in self.state.board if any(cell == 0 for cell in row)]
        cleared = GRID_HEIGHT - len(new_board)
        self.state.board = [[0] * GRID_WIDTH for _ in range(cleared)] + new_board
        self.state.score += cleared * 100

    def _try_rotate(self, new_rot: int):
        shape = self.state.current_piece_shape
        if shape == 'O':
            self.state.rot = new_rot
            return

        kicks = (SRS_I if shape == 'I' else SRS_JLSTZ).get(
            (self.state.rot, new_rot), [(0, 0)]
        )
        new_piece = MINOS[shape][new_rot]
        old_pos = self.state.current_piece_position[:]

        for dx, dy in kicks:
            test_pos = [old_pos[0] - dy, old_pos[1] + dx]
            if self._is_valid_position(new_piece, test_pos):
                self.state.current_piece = new_piece
                self.state.current_piece_position = test_pos
                self.state.rot = new_rot
                self._update_ghost()
                return

    # --- public actions ---

    def move_left(self):
        self.state.current_piece_position[1] -= 1
        if not self._is_valid_position():
            self.state.current_piece_position[1] += 1
        else:
            self._update_ghost()

    def move_right(self):
        self.state.current_piece_position[1] += 1
        if not self._is_valid_position():
            self.state.current_piece_position[1] -= 1
        else:
            self._update_ghost()

    def move_down(self):
        self.state.current_piece_position[0] += 1
        if not self._is_valid_position():
            self.state.current_piece_position[0] -= 1
            self._lock_piece()
            self._clear_lines()
            self._pop_next_piece()

    def hard_drop(self):
        self.state.current_piece_position = self.state.ghost_position[:]
        self._lock_piece()
        self._clear_lines()
        self._pop_next_piece()

    def hold(self):
        if self.state.hold_piece_shape is None:
            self.state.hold_piece_shape = self.state.current_piece_shape
            self._pop_next_piece()
        else:
            self.state.current_piece_shape, self.state.hold_piece_shape = (
                self.state.hold_piece_shape,
                self.state.current_piece_shape,
            )
            self.state.rot = 0
            self.state.current_piece = MINOS[self.state.current_piece_shape][0]
            self.state.current_piece_position = [0, GRID_WIDTH // 2 - 2]
            self._update_ghost()

    def rotate_clockwise(self):
        self._try_rotate((self.state.rot + 1) % 4)

    def rotate_counterclockwise(self):
        self._try_rotate((self.state.rot - 1) % 4)

    def gravity_tick(self):
        self.move_down()

    def get_state(self) -> dict:
        return self.state.to_dict()

File: tetris-app/frontend/src/game/gameController.ts
import { PIECES, INITIAL_PIECE_POSITION, PieceShape, GRID_WIDTH, GRID_HEIGHT, SRS_JLSTZ, SRS_I, MINOS, NUM_NEXT_PIECES, SRS_180 } from "./pieces"
import { Board, Position, initialGameState } from "./gameState"
import { GameState } from "./../hooks/useGame"
import { shuffle } from "lodash"

type ClearLinesResult = { board: Board, clearedCount: number }

// --- internal helpers ---

export function isValidPosition(piece: PieceShape, pos: Position, board: Board): boolean {
  for (let y = 0; y < piece.length; y++) {
    for (let x = 0; x < piece[y].length; x++) {
      const cell = piece[y][x]
      if (cell) {
        const nx = pos[1] + x
        const ny = pos[0] + y
        if (nx < 0 || nx >= GRID_WIDTH || ny >= GRID_HEIGHT) {
          return false
        }
        if (ny >= 0 && board[ny][nx]) {
          return false
        }
      }
    }
  }
  return true
}

export function updateGhost(piece: PieceShape, pos: Position, board: Board): Position {
  const ghostpos: Position = [...pos]
  while (isValidPosition(piece, ghostpos, board)) {
    ghostpos[0] += 1
  }
  ghostpos[0] -= 1
  return ghostpos
}

export function lockPiece(state: GameState): GameState {
  const shape = state.current_piece_shape
  const pos = state.current_piece_position
  const piece = state.current_piece

  const newBoard = state.board.map(row => [...row])

  for (let y = 0; y < piece.length; y++) {
    for (let x = 0; x < piece[y].length; x++) {
      const cell = piece[y][x]
      if (cell) {
        newBoard[pos[0] + y][pos[1] + x] = shape
      }
    }
  }

  return { ...state, board: newBoard }
}

export function clearLines(state: GameState): ClearLinesResult {
  let newBoard = state.board.filter((row) => row.some(cell => cell === 0))
  const cleared = GRID_HEIGHT - newBoard.length
  newBoard = [...new Array(cleared).fill(0).map(() => new Array(10).fill(0)), ...newBoard]
  return { board: newBoard, clearedCount: cleared }
}

// SRSに基づいて回転する
function tryRotate(state: GameState, newRot: number): GameState | null {
  const board = state.board
  const shape = state.current_piece_shape
  const newPiece = MINOS[shape][newRot]
  const oldPos = state.current_piece_position

  if (shape === "O") {
    return { ...state, rot: newRot }
  }

  const kicks = shape === "I" ? SRS_I : SRS_JLSTZ
  const key = state.rot + "," + newRot
  const kickList = kicks[key] ?? [[0, 0]]

  for (const [dx, dy] of kickList) {
    const testPos: Position = [oldPos[0] - dy, oldPos[1] + dx]
    if (isValidPosition(newPiece, testPos, board)) {
      const newGhost = updateGhost(newPiece, testPos, board)
      return {
        ...state,
        current_piece: newPiece,
        current_piece_position: testPos,
        rot: newRot,
        ghost_position: newGhost,
      }
    }
  }
  return null
}

// ネクストの補充（毎回ランダムにして入れている。50手ぐらい先まで入れておいてもいいかも？）
export function refillNext(nextPieces: string[]): string[] {
  if (nextPieces.length <= NUM_NEXT_PIECES) {
    const bag = shuffle([...PIECES])
    return [...nextPieces, ...bag]
  }
  return nextPieces
}

// ネクストの先頭を取り出して、次のミノをセットする
export function popNextPiece(state: GameState): GameState {
  let gameOver = false
  const refilled = refillNext(state.next_pieces)
  const newRot: number = 0
  const [newPieceShape, ...nextPieces] = refilled

  // 分割代入の結果が string | undefined のため
  if (newPieceShape === undefined) {
    throw new Error("next_pieces が空です。refillNext が正しく動いていない可能性があります")
  }

  const newPiece = MINOS[newPieceShape][newRot]
  const newPos: Position = [...INITIAL_PIECE_POSITION]
  const newGhost = updateGhost(newPiece, newPos, state.board)

  if (!isValidPosition(newPiece, newPos, state.board)) {
    gameOver = true
  }

  return {
    ...state,
    current_piece_shape: newPieceShape,
    current_piece: newPiece,
    current_piece_position: newPos,
    ghost_position: newGhost,
    rot: newRot,
    next_pieces: nextPieces,
    game_over: gameOver,
  }
}

// --- public actions ---

export function isGrounded(state: GameState): boolean {
  const testPos: Position = [state.current_piece_position[0] + 1, state.current_piece_position[1]]
  return !isValidPosition(state.current_piece, testPos, state.board)
}

export function forceLock(state: GameState): GameState {
  let newState = lockPiece(state)
  const { board, clearedCount } = clearLines(newState)
  newState = { ...newState, board: board }
  newState = popNextPiece(newState)
  return newState
}

export function moveLeft(state: GameState): GameState | null {
  const board = state.board
  const piece = state.current_piece
  const testPos: Position = [...state.current_piece_position]
  testPos[1] -= 1

  if (!isValidPosition(piece, testPos, board)) {
    return null
  }
  const newGhostPos = updateGhost(piece, testPos, board)
  return {
    ...state,
    current_piece_position: testPos,
    ghost_position: newGhostPos,
  }
}

export function moveRight(state: GameState): GameState | null {
  const board = state.board
  const piece = state.current_piece
  const testPos: Position = [...state.current_piece_position]
  testPos[1] += 1

  if (!isValidPosition(piece, testPos, board)) {
    return null
  }
  const newGhostPos = updateGhost(piece, testPos, board)
  return {
    ...state,
    current_piece_position: testPos,
    ghost_position: newGhostPos,
  }
}

export function moveDown(state: GameState): GameState | null {
  const board = state.board
  const piece = state.current_piece
  const testPos: Position = [...state.current_piece_position]
  testPos[0] += 1
  if (!isValidPosition(piece, testPos, board)) {
    return null
  }
  return {
    ...state,
    current_piece_position: testPos,
  }
}

// 今のミノを下に落として固定する
export function hardDrop(state: GameState): GameState {
  const dropPos = state.ghost_position
  let newState = lockPiece({ ...state, current_piece_position: dropPos })
  const { board, clearedCount } = clearLines(newState)
  newState = { ...newState, board: board }
  newState = popNextPiece(newState)
  return newState
}

// ホールド機能
export function hold(state: GameState): GameState {
  if (state.hold_piece_shape === null) {
    const newState = { ...state, hold_piece_shape: state.current_piece_shape }
    return popNextPiece(newState)
  } else {
    const newPieceShape = state.hold_piece_shape
    const newPiece = MINOS[newPieceShape][0]
    const newPos: Position = [...INITIAL_PIECE_POSITION]
    const newGhost = updateGhost(newPiece, newPos, state.board)
    return {
      ...state,
      current_piece_shape: newPieceShape,
      current_piece: newPiece,
      current_piece_position: newPos,
      ghost_position: newGhost,
      rot: 0,
      hold_piece_shape: state.current_piece_shape,
    }
  }
}

export function rotateClockwise(state: GameState): GameState | null {
  const newRot = (state.rot + 1) % 4
  return tryRotate(state, newRot)
}

export function rotateCounterClockwise(state: GameState): GameState | null {
  const newRot = (state.rot - 1 + 4) % 4
  return tryRotate(state, newRot)
}

// 180度回転はSRSに含まれていないため、独自に実装する（面倒な実装）
export function rotate180(state: GameState): GameState | null {
  const shape = state.current_piece_shape
  if (shape === "O") {
    return null
  }

  const newRot = (state.rot + 2) % 4
  const newPiece = MINOS[shape][newRot]
  const oldPos = state.current_piece_position

  for (const [dx, dy] of SRS_180) {
    const testPos: Position = [oldPos[0] - dy, oldPos[1] + dx]
    if (isValidPosition(newPiece, testPos, state.board)) {
      const newGhost = updateGhost(newPiece, testPos, state.board)
      return {
        ...state,
        current_piece: newPiece,
        current_piece_position: testPos,
        rot: newRot,
        ghost_position: newGhost,
      }
    }
  }
  return null
}

export function gravityTick(state: GameState): GameState | null {
  return moveDown(state)
}

export function restart(): GameState {
  let newState = { ...initialGameState }
  newState = {
    ...newState,
    next_pieces: refillNext(newState.next_pieces),
  }
  newState = popNextPiece(newState)
  return newState
}